src/avram/polymorphic.cr

```diff
@@ -0,0 +1,144 @@
+# Helpers for defining polymorphic associations
+module Avram::Polymorphic
+  # Sets up helper methods for validating and loading polymorphic associatiions
+  #
+  # This will generate methods for validiating, preloading, and lazy loading
+  # polymorphic associations.
+  #
+  # Let's start with this example:
+  #
+  # > *Important note:* the `belongs_to` associations must be nilable since only
+  # > one polymorphic can be set at a time. Remember to also make the association
+  # > nilable in the migration, for example: `add_belongs_to photo : Photo?`
+  #
+  # ```
+  # class Comment < BaseModel
+  #   table do
+  #     belongs_to photo : Photo?
+  #     belongs_to video : Video?
+  #     polymorphic :commentable, associations: [:photo, :video]
+  #   end
+  # end
+  #
+  # class Photo < BaseModel
+  #   table do
+  #     has_many comments : Comment
+  #   end
+  # end
+  #
+  # class Photo < BaseModel
+  #   table do
+  #     has_many comments : Comment
+  #   end
+  # end
+  # ```
+  #
+  # `has_many` calls work just like normal and so do `belongs_to`.
+  #
+  # `polymorphic` just set up methods for validation, preloading, and getting
+  # the polymorphic association more easily.
+  #
+  # ## Validations
+  #
+  # `polymorphic` will add a validation so that exactly one of the associations
+  # is present. In this case either `photo_id` or `video_id`, but not both.
+  #
+  # If `optional` is set to true, it will allow the associations to be `nil`, but
+  # will still not allow more than one association to be set at a time.
+  #
+  # ## Preloading
+  #
+  # `polymorphic` defines a `preload_{polymorphic_name}` on the query.
+  #
+  # In this example it would define `preload_commentable` on `Comment::BaseQuery`.
+  # Under the hood this is preloading each belongs_to. In this example
+  # `Comment::BaseQuery.preload_commentable` is like calling
+  # `Comment::BaseQuery.preload_photo.preload_video`.
+  #
+  # ## Model getter
+  #
+  # Adds `{polymorphic_name}` and `{polymorphic_name}!` getters to the model.
+  #
+  # In this example `commentable` and `commentable!`. Calling these will return
+  # either the `video` or the `photo`. The `!` version will allow lazy loading.
+  # The non `!` version requires preloading with `preload_{polymorphic_name}.
+  #
+  # If `optional` is set to true these methods will return `nil` if nothing
+  # is associated.
+  #
+  # ```
+  # Comment::BaseQuery.new.each do |comment|
+  #   comment.commentable! # Returns the photo or video
```

Comment:
I _think_ we'd want them to preload explicitly more often than not, yeah? If so, I'd change this example to use `preload_commentable` and the non-bang method

Reply:
Yeah that’s a good point. I was going for concise but that’s probably a better idea.